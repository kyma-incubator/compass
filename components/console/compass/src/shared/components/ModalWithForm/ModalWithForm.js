import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import { ControlledModal } from '../Modal/ControlledModal';
import { Button } from 'fundamental-react/Button';
import LuigiClient from '@luigi-project/client';
import { useMutationObserver } from 'react-shared';
import { SEND_NOTIFICATION } from 'gql';
import { useMutation } from 'react-apollo';

const ModalWithForm = ({
  performRefetch,
  title,
  button,
  confirmText,
  initialIsValid,
  renderForm,
  modalClassName,
}) => {
  const [isOpen, setOpen] = useState(false);
  const [isValid, setValid] = useState(initialIsValid);
  const [customValid, setCustomValid] = useState(true);
  const [sendNotification] = useMutation(SEND_NOTIFICATION);
  const formElementRef = useRef(null);

  const handleFormChanged = e => {
    setValid(formElementRef.current.checkValidity());
    if (typeof e.target.reportValidity === 'function') {
      // for IE
      e.target.reportValidity();
    }
  };

  useMutationObserver(formElementRef, () => {
    if (formElementRef.current) {
      handleFormChanged({ target: formElementRef.current });
    }
  });

  const setOpenStatus = status => {
    if (status) {
      LuigiClient.uxManager().addBackdrop();
    } else {
      LuigiClient.uxManager().removeBackdrop();
    }
    setOpen(status);
  };

  const handleFormError = (title, message) => {
    sendNotification({
      variables: {
        content: message,
        title: title,
        color: '#BB0000',
        icon: 'decline',
      },
    });
  };

  const handleFormSuccess = (title, message) => {
    sendNotification({
      variables: {
        content: message,
        title: title,
        color: '#107E3E',
        icon: 'accept',
      },
    });

    performRefetch();
  };

  const onConfirm = () => {
    const form = formElementRef.current;
    if (
      typeof form.reportValidity === 'function'
        ? form.reportValidity()
        : form.checkValidity() // IE workaround; HTML validation tooltips won't be visible
    ) {
      setOpenStatus(false);
      form.dispatchEvent(new Event('submit'));
    }
  };

  const actions = (
    <>
      <Button
        option="light"
        onClick={() => setOpenStatus(false)}
        className="fd-has-margin-right-s"
      >
        Cancel
      </Button>
      <Button
        aria-disabled={!isValid || !customValid}
        onClick={onConfirm}
        option="emphasized"
      >
        {confirmText}
      </Button>
    </>
  );

  const modalOpeningComponent = (
    <Button
      glyph={button.glyph || null}
      option={button.option}
      onClick={() => setOpenStatus(true)}
    >
      {button.text}
    </Button>
  );

  return (
    <ControlledModal
      modalOpeningComponent={modalOpeningComponent}
      modalClassName={modalClassName}
      show={isOpen}
      actions={actions}
      title={title}
      onClose={() => setOpenStatus(false)}
    >
      {renderForm({
        formElementRef,
        isValid,
        setCustomValid: isValid => {
          // revalidate rest of the form
          setValid(formElementRef.current.checkValidity());
          setCustomValid(isValid);
        },
        onChange: handleFormChanged,
        onError: handleFormError,
        onCompleted: handleFormSuccess,
      })}
    </ControlledModal>
  );
};

ModalWithForm.propTypes = {
  performRefetch: PropTypes.func.isRequired,
  sendNotification: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  confirmText: PropTypes.string.isRequired,
  initialIsValid: PropTypes.bool,
  button: PropTypes.exact({
    text: PropTypes.string.isRequired,
    glyph: PropTypes.string,
    option: PropTypes.oneOf(['emphasized', 'light']),
  }).isRequired,
  modalClassName: PropTypes.string,
  renderForm: PropTypes.func.isRequired,
};
ModalWithForm.defaultProps = {
  sendNotification: () => {},
  performRefetch: () => {},
  initialIsValid: false,
};

export default ModalWithForm;
