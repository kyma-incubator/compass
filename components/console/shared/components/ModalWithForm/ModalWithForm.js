import React, { useState, useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import { Modal, Button } from 'fundamental-react';
import LuigiClient from '@luigi-project/client';
import { useNotification } from '../../contexts/NotificationContext';
import { Tooltip } from '../Tooltip/Tooltip';
import CustomPropTypes from '../../typechecking/CustomPropTypes';

const isFormValid = (formRef, reportValidity = false) => {
  if (!formRef || !formRef.current) return true;

  if (reportValidity && typeof formRef.current.reportValidity === 'function') {
    // for IE
    formRef.current.reportValidity();
  }

  return formRef.current.checkValidity();
};

const isJsonSchemaFormValid = formRef => {
  if (!formRef || !formRef.current) return true;

  return formRef.current.state && formRef.current.state.errors
    ? !formRef.current.state.errors.length
    : true;
};

export const ModalWithForm = ({
  performRefetch,
  sendNotification,
  title,
  button,
  renderForm,
  opened,
  customCloseAction,
  item,
  modalOpeningComponent,
  confirmText,
  invalidPopupMessage,
  ...props
}) => {
  const [isOpen, setOpen] = useState(false);
  const [isValid, setValid] = useState(false);
  const [customValid, setCustomValid] = useState(true);
  const formElementRef = useRef(null);
  const jsonSchemaFormRef = useRef(null);
  const notificationManager = useNotification();

  useEffect(() => {
    setOpenStatus(opened);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [opened]);

  function checkAllForms(reportValidity = false) {
    const _isEveryFormValid =
      isFormValid(formElementRef, reportValidity) &&
      isJsonSchemaFormValid(jsonSchemaFormRef);
    if (isValid !== _isEveryFormValid) {
      setValid(_isEveryFormValid);
    }
  }

  // useEffect(() => {
  //   setTimeout(() => checkAllForms(true));
  // });

  function setOpenStatus(status) {
    if (status) {
      LuigiClient.uxManager().addBackdrop();
    } else {
      LuigiClient.uxManager().removeBackdrop();
      if (customCloseAction) {
        customCloseAction();
      }
    }
    setOpen(status);
  }

  function handleFormChanged(e) {
    setTimeout(() => checkAllForms());
    if (e.target) {
      if (e.target.getAttribute('data-ignore-visual-validation')) {
        return;
      }

      // current element validity
      const isValidFormElement = e.target.checkValidity();
      e.target.classList.toggle('is-invalid', !isValidFormElement);
    }
  }

  function handleFormError(title, message, isWarning) {
    notificationManager.notifyError({
      content: message,
      title: title,
      type: isWarning ? 'warning' : 'error',
    });
  }

  function handleFormSuccess(title, message) {
    // notificationManager.notify({
    //   content: message,
    //   title: title,
    //   color: '#107E3E',
    //   icon: 'accept',
    //   autoClose: true,
    // });

    performRefetch();
  }

  function handleFormSubmit() {
    const _isEveryFormValid =
      isFormValid(formElementRef) && isJsonSchemaFormValid(jsonSchemaFormRef);
    if (_isEveryFormValid) {
      formElementRef.current.dispatchEvent(new Event('submit'));
      setTimeout(() => setOpenStatus(false));
    }
  }

  function renderConfirmButton() {
    const disabled = !isValid || !customValid;
    const button = (
      <Button
        disabled={disabled}
        aria-disabled={disabled}
        onClick={handleFormSubmit}
        option="emphasized"
      >
        {confirmText}
      </Button>
    );

    if (invalidPopupMessage && disabled) {
      return (
        <Tooltip
          content={invalidPopupMessage}
          position="top"
          trigger="mouseenter"
          tippyProps={{
            distance: 16,
          }}
        >
          {button}
        </Tooltip>
      );
    }
    return button;
  }

  return (
    <>
      {modalOpeningComponent ? (
        <div
          style={{ display: 'contents' }}
          onClick={() => setOpenStatus(true)}
        >
          {modalOpeningComponent}
        </div>
      ) : (
        <Button
          glyph={button.glyph || null}
          aria-label={button.label || null}
          option={button.option}
          compact={button.compact || false}
          disabled={!!button.disabled}
          onClick={() => setOpenStatus(true)}
        >
          {button.text}
        </Button>
      )}
      <Modal
        {...props}
        show={isOpen}
        actions={
          <>
            <Button
              onClick={() => {
                setOpenStatus(false);
              }}
              option="light"
            >
              Cancel
            </Button>
            {renderConfirmButton()}
          </>
        }
        onClose={() => {
          setOpenStatus(false);
        }}
        title={title}
      >
        {renderForm({
          formElementRef,
          jsonSchemaFormRef,
          isValid,
          setCustomValid: isValid => {
            // revalidate rest of the form
            setValid(formElementRef.current.checkValidity());
            setCustomValid(isValid);
          },
          onChange: handleFormChanged,
          onError: handleFormError,
          onCompleted: handleFormSuccess,
          performManualSubmit: handleFormSubmit,
          item: item,
        })}
      </Modal>
    </>
  );
};

ModalWithForm.propTypes = {
  performRefetch: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  renderForm: PropTypes.func.isRequired,
  opened: PropTypes.bool,
  customCloseAction: PropTypes.func,
  item: PropTypes.object,
  modalOpeningComponent: PropTypes.node,
  confirmText: PropTypes.string,
  invalidPopupMessage: PropTypes.string,
  button: CustomPropTypes.button,
};

ModalWithForm.defaultProps = {
  performRefetch: () => {},
  confirmText: 'Create',
  invalidPopupMessage: '',
};
