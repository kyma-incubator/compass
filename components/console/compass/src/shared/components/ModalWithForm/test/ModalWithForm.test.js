import React from 'react';
import renderer from 'react-test-renderer';
import ModalWithForm from '../ModalWithForm';

describe('ModalWithForm', () => {
  it('Renders child component', () => {
    const child = <span>test</span>;
    const component = renderer.create(
      <ModalWithForm
        title=""
        performRefetch={() => {}}
        sendNotification={() => {}}
        confirmText="Create"
        button={{ text: '' }}
        renderForm={() => child}
      />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
