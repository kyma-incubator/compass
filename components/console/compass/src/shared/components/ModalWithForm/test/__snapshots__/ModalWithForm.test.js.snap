// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalWithForm Renders child component 1`] = `
<div>
  <button
    className="fd-button"
    onClick={[Function]}
  >
    
  </button>
</div>
`;
